### ISSUE TITLE
Pared down boilerplate

### ISSUE THREAD

**Issue by dsmithhfx** [ngbp/ngbp#160]
In the readme/Roadmap section you mention the potential enhancement of a 'custom download builder'... until that happy day arrives, can you suggest guidelines to help ensure the success of removing these components from ng-boilerplate:
LESS
Bootstrap
AngularUI Bootstrap
Font Awesome
-- so that the out-of-the-box build continues to function on some level?

These components are certainly worthy for Bootstrap-based projects, but could prove highly distracting for non-Bootstrap projects. And LESS appears to depend on a node.js server or 'modern' browser, both of which are likely deal-breakers for many hosting and end-use scenarios.

For all I know, removal of these components may well be very simple and straightforward, but just thought I'd ask first in case there are references buried in build scripts that need to be edited.

Thanks!

**Comment by joshdmiller** [comment]
Hey @dsmithhfx. That day is close upon us. :-)

Version 0.4, which is due out in the next couple of days, has a crazy-simple way of enabling and disabling optional components, right from `package.json`. There is also a grunt interface to it:

``` sh
grunt ngbp:enable:less
grunt ngbp:disable:jade
```

Part of that idea, too, is splitting the demo app from the landing page / marketing site. In other words, when someone downloads ngBoilerplate to start their project, they won't have _any_ optional or vendor components installed, such as less and bootstrap.

Within the next few weeks I also hope to have a primitive CLI tool that downloads _only_ the build script so completely bare projects can also be boostrapped with ngBoilerplate.

The work for this is on the `modularization-wip` branch: https://github.com/ngbp/ng-boilerplate/tree/modularization-wip. If you have ideas, I'd love to hear them - this is all very experimental.

**Comment by dsmithhfx** [comment]
Great news! I look forward to the new version, and maybe once I've spent some time with it (and got my head around Angular), I'll have some ideas. Thanks.

**Comment by rosslavery** [comment]
Hey John, is part of 0.4 going to include the ability to switch between LESS and SASS/Compass?

In https://github.com/ngbp/ng-boilerplate/issues/31 you mentioned you had a solution, just not sure if this is referring to the same thing.

Thanks for all your work on this.

**Comment by joshdmiller** [comment]
@rabhw Indeed it will. :-)

The possible solution I mentioned in #31 has been superceeded by a _far_ superior method, but the same idea is there. To use sass, you'll simply issue:

``` sh
$ grunt ngbp:enable:sass
```

And it will install the correct NPM modules and load the task into the correct places in the build process. No further effort necessary. It's experimental, but working well so far; it should be pretty cool.

**Comment by dsmithhfx** [comment]
@rabhw 
there's a somewhat similar project/tutorial that installs sass by default described here:
http://newtriks.com/2013/06/11/automating-angularjs-with-yeoman-grunt-and-bower/
@joshdmiller 
I'm confused about LESS -- does it (or can it) compile to 'plain vanilla' css, so production code need not rely on node server or 'modern browser'?

**Comment by joshdmiller** [comment]
LESS and SASS are very similar; they are both CSS supersets and "compile down" to regular old CSS. The resulting CSS is indistinguishable from CSS written by other means. While it's _possible_ to run less directly in the browser if you include its accompanying JavaScript file, I've never seen this done in production.

If you want to learn more about the philsophical differences between ngBoilerplate and the AngularJS generators for Yeoman, see yeoman/generator-angular#109.

**Comment by rosslavery** [comment]
@joshdmiller That sounds amazing. Very much looking forward to the next release.

@dsmithhfx My project is currently in yeoman but I'll be migrating to ngbp as soon as 0.4 is out. Josh covers many of the reasons for my migration in the post he linked. Having experienced the project structure of yeoman, I've learned that I much prefer my modules to be self-contained and adjacent to all of their related files, rather than being spread across a variety of folders (controllers, directives, services, etc.)

**Comment by dsmithhfx** [comment]
@joshdmiller I'm not sure where I saw something similar to this, maybe it was here (but I don't think so):
https://npmjs.org/package/generator-cg-angular
"The project will include a ready-made Grunt build that will:
    _Build all the LESS files into one minified CSS file._
    [...]"
So LESS can be still be used for development without imposing burdensome app run time requirements.

**Comment by joshdmiller** [comment]
@dsmithhfx I'm a little confused - what issue are you trying to solve? LESS doesn't introduce _any_ runtime requirements. ngBoilerplate - like every other development stack - runs the less toolchain to convert LESS to CSS before the developer ever loads it onto their server - or even, in the case of ngBoilerplate, previews it during development. So when you're running from a development build or running in production, all that's ever used is CSS. Maybe I'm missing what you're asking...

**Comment by dsmithhfx** [comment]
@joshdmiller Yeah, sorry if my question wasn't clear, I just needed to do a little digging to find the answer. The deployment scenarios I typically face are pretty constrained, so that was my major concern (unfounded as it turned out) with LESS etc.

**Comment by Keats** [comment]
Making your own build is not that hard, I wanted to do the same things as you said and ended up writing my own version of ng-boilerplate (https://github.com/Keats/ng-boilerplate) for personal use.  
Once you understand Grunt, it is pretty straightforward (especially when you don't need/want to add customisations)

**Comment by cyrusdavid** [comment]
:+1: The boilerplate got bloated up with those components. But it still is an awesome boilerplate! :P

**Comment by joshdmiller** [comment]
v0.4 is ultra-minimalist, containing _no_ Bower packages (save AngularJS, of course!) out of the box. Release in a couple weeks; developer preview end of next week. Anyone interested in testing the alpha end of next week?

**Comment by benshaw** [comment]
@joshdmiller any eta on v0.4 ? In particular im looking for SASS support. thanks.

**Comment by vivekxp** [comment]
Ya, I have been waiting for v0.4 for my next project too. Any release plan? I am interested in sass too.

**Comment by WillsB3** [comment]
+1

**Comment by joshhornby** [comment]
Any update on the Sass support @joshdmiller

**Comment by blowsie** [comment]
Is this project still alive? The comment below is over a year old and the project is still 0.3?

> Version 0.4, which is due out in the next couple of days

**Comment by Merott** [comment]
@blowsie 

The ngbp project in particular is not going to receive new features, though good quality PRs that fix existing issues will be accepted.

See the conversation on [this thread](https://github.com/ngbp/ngbp/issues/330) for more on what's going on.
